File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        queue = collections.deque([(root, float("-inf"), float("inf"))])

        while queue:
            node, leftMin, rightMax = queue.popleft()
            if node:
                if not (leftMin < node.val < rightMax):
                    return False
                queue.append((node.left, leftMin, node.val))
                queue.append((node.right, node.val, rightMax))
        return True


            

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private List<Integer> nums;
    public Solution() {
        this.nums = new ArrayList<>();
    }

    public void dfs(TreeNode root) {
        if(root == null) return;
        dfs(root.left);
        nums.add(root.val);
        dfs(root.right);      
    }

    public int kthSmallest(TreeNode root, int k) {
        dfs(root);
        return nums.get(k - 1);
    }
}
